esmvalcore/cmor/_fixes/icon/_base_fixes.py

```diff
@@ -58,47 +208,105 @@ def get_horizontal_grid(self, cube):
-                f"Cube does not contain the attribute '{self.GRID_FILE_ATTR}' "
-                f"necessary to download the ICON horizontal grid file:\n"
-                f"{cube}")
-        grid_url = cube.attributes[self.GRID_FILE_ATTR]
+        (grid_url, grid_name) = self._get_grid_url(cube)
 
         # If already loaded, return the horizontal grid (cube)
-        parsed_url = urlparse(grid_url)
-        grid_name = Path(parsed_url.path).name
         if grid_name in self._horizontal_grids:
-            return self._horizontal_grids[grid_name]
+            return self._horizontal_grids[grid_name].copy()
 
         # Check if grid file has recently been downloaded and load it if
         # possible
-        grid_path = self.CACHE_DIR / grid_name
-        if grid_path.exists():
-            mtime = grid_path.stat().st_mtime
-            now = datetime.now().timestamp()
-            age = now - mtime
-            if age < self.CACHE_VALIDITY:
-                logger.debug("Using cached ICON grid file '%s'", grid_path)
-                self._horizontal_grids[grid_name] = self._load_cubes(grid_path)
-                return self._horizontal_grids[grid_name]
-            logger.debug("Existing cached ICON grid file '%s' is outdated",
-                         grid_path)
-
-        # Download file if necessary
-        logger.debug("Attempting to download ICON grid file from '%s' to '%s'",
-                     grid_url, grid_path)
-        self.CACHE_DIR.mkdir(parents=True, exist_ok=True)
-        with requests.get(grid_url, stream=True,
-                          timeout=self.TIMEOUT) as response:
-            response.raise_for_status()
-            with open(grid_path, 'wb') as file:
-                copyfileobj(response.raw, file)
-        logger.info("Successfully downloaded ICON grid file from '%s' to '%s'",
-                    grid_url, grid_path)
-
-        self._horizontal_grids[grid_name] = self._load_cubes(grid_path)
-        return self._horizontal_grids[grid_name]
+        # Note: we use a lock here to prevent multiple processes from
+        # downloading the file simultaneously and to ensure that other
+        # processes wait until the download has finished
+        lock = FileLock(self.CACHE_DIR / f"{grid_name}.lock")
+        with lock:
+            grid_path = self.CACHE_DIR / grid_name
+            if grid_path.exists():
+                mtime = grid_path.stat().st_mtime
+                now = datetime.now().timestamp()
+                age = now - mtime
+                if age < self.CACHE_VALIDITY:
+                    logger.debug("Using cached ICON grid file '%s'", grid_path)
+                    self._horizontal_grids[grid_name] = self._load_cubes(
+                        grid_path
+                    )
+                    return self._horizontal_grids[grid_name].copy()
+                logger.debug("Existing cached ICON grid file '%s' is outdated",
+                             grid_path)
```

Comment:
this whole block should probably sit outside the lock context manager - you are just testing for the age of the file, and only initiate the locked download if the file is young enough

Reply:
I thought so too at the beginning, but consider the following scenario (which I encountered in practice).

Process 1 starts the download and creates the grid file. Process 2 sees that the file exists and tries to use it, but since the download has not finished the file cannot be loaded --> Error.

Thus, the `if grid_path.exists():` needs to be inside the context manager. I would need a check for the existence of the file within the lock anyway, otherwise locked processes would still initiate the download after the lock has been released.

Reply:
but are you checking anywhere for the existence of the `lock` file? Coz that's the one you don't want to have lingering about. At any rate, gonna approve now since it looks to be running fine and it actually dies from you guys' tests (not tested myself) :+1:

Reply:
No, I am not, but that would be `filelock`'s job in the context manager :smile: However, the lock does not rely on the existence of the file, but rather uses a OS-related lock (see [here](https://github.com/tox-dev/py-filelock/blob/d8c6eb077b9695b0886ff9c10945a2687d245f09/src/filelock/_unix.py#L33-L41) for UNIX). That also means that the lock file persists after successful runs (but since this is a cache directory I don't see a problem with that).

Reply:
hmm I thought the `lock` file needs to go poof when the locked operations finished successfully - maybe I'm too oldschool and things changed forwards in Filelock since I've last used it

Reply:
Yeah I though so too, but as you can see in the source code of `filelock` above, acquiring/releasing the lock does not create/delete the lock file, but uses [`fcntl.flock`](https://docs.python.org/3/library/fcntl.html#fcntl.flock) (if the file is not present it of course creates it first).

Reply:
I'm pressing the disengage button before I start concerning myself with flocks and glocks :rofl: